File: NSP/Sources/calculus.c
//
//  calculus.c
//  MUTE
//
//  Created by Kota on 11/22/R6.
//
#include"calculus.h"
__attribute__((always_inline))
void calculus_integration(register double const * __nonnull x,
						  register double const * __nonnull y, // reset
						  register double       * __nonnull z,
						  simd_double2 * __nonnull const w,
						  intptr_t const length) {
	register simd_double2 s = *w;
	for ( register double const * __nonnull const _ = z + length ; z < _ ; ++ x, ++ y, ++ z ) {
		*z = (s = (simd_double2 const) {
			*x,
			(*y<=0) * simd_dot((simd_double3 const){0.5, 1, 0.5}, (simd_double3 const){s.x, s.y, *x})
		}).y;
	}
	*w = s;
}
__attribute__((always_inline))
void calculus_differentiation(register double const * __nonnull x,
							  register double const * __nonnull y, // reset
							  register double       * __nonnull z,
							  simd_double2 * __nonnull const w,
							  intptr_t const length) {
	register simd_double2 s = *w;
	for ( register double const * __nonnull const _ = z + length ; z < _ ; ++ x, ++ y, ++ z )
		*z = (s = (simd_double2 const) {
			*x,
			simd_dot((simd_double3 const){-2, (0<*y) - 1, 2}, (simd_double3 const){s.x, s.y, *x}),
		}).y;
	*w = s;
}

File: NSP/Tests/Calculus.swift
//
//  Calculus.swift
//  MUTE
//
//  Created by Kota on 11/22/R6.
//
import Testing
import NSP
@Suite
struct Calculus {
	@Test
	func integration() {
		let z = Array<Float64>(repeating: 0, count: 8)
		let x = (0..<8).map(Float64.init)
		var y = Array<Float64>(repeating: 0, count: 8)
		var w = SIMD2<Float64>(0, 0)
		calculus_integration(x, z, &y, &w, 8)
		print(y)
	}
	
	@Test
	func differentiation() {
		let z = Array<Float64>(repeating: 0, count: 8)
		let y = (0..<8).map(Float64.init)//(0..<8).map { Float64($0*$0) * 0.5 }
		var x = Array<Float64>(repeating: 0, count: 8)
		var w = SIMD2<Float64>(0, 0)
		calculus_differentiation(y, z, &x, &w, 8)
		print(x)
	}
}
